storage/redis.js

```diff
@@ -0,0 +1,343 @@
+      return []
+    }
+  }
+
+  /**
+   * @param {string} name
+   */
+  async clear(name) {
+    this.log.debug({ msg: 'acd/storage/redis.clear', name })
+
+    try {
+      if (!name) {
+        await this.store.flushall()
+        return
+      }
+
+      const keys = await this.store.keys(`${name}*`)
+      this.log.debug({ msg: 'acd/storage/redis.clear keys', keys })
+
+      const removes = keys.map(key => ['del', key])
+      await this.store.pipeline(removes).exec()
+
+      if (!this.invalidation) { return }
+      await this.clearReferences(keys)
+    } catch (err) {
+      this.log.error({ msg: 'acd/storage/redis.clear error', err, name })
+    }
+  }
+
+  async refresh() {
+    try {
+      await this.store.flushall()
+    } catch (err) {
+      this.log.error({ msg: 'acd/storage/redis.refresh error', err })
+    }
+  }
+
+  /**
+   * TODO DOC!!!
+   * options.invalidation.clear.minInterval min interval between two clear operations, default 1 sec
+   * options.invalidation.clear.maxInterval min interval between two clear operations, default 10 sec
+   * options.invalidation.clear.maxTime heuristic max time to full clear the dangling references
+   */
+  heuristicClean({ start = false, interval, lastCursor, retrieves = 10 } = {}) {
+    this.log.debug({ msg: 'acd/storage/redis.heuristicClean' })
+
+    if (start) {
+      this._cleanTimer = setTimeout(() => { this.heuristicClean() }, this.options.invalidation.clear.minInterval)
+      return
+    }
+
+    // get keys->references count
+    const [cursor, keys] = await this.store.sscan(lastCursor || 0, 'match', 'k:*', 'count', retrieves)
+
+    this.log.debug({ msg: 'acd/storage/redis.heuristicClean sscan', cursor, keys })
+
+    if(keys.length < 1) {
+      setTimeout(() => { this.heuristicClean() }, this.options.invalidation.clear.minInterval)
+    }
+
+    if (cursor === 0 && lastCursor) {
+      // made a "full iteration", so it means the keys count
+      // see https://redis.io/commands/scan
+      // assuming all keys will expire in maxTime
+      interval = (lastCursor + keys.length) / this.options.invalidation.clear.maxTime
+      // cap interval
+      interval = Math.max(this.options.invalidation.clear.minInterval, Math.min(this.options.invalidation.clear.maxInterval, interval))
+    }
+
+    let key = randomElement(keys)
+    key = key.substr(2) // remove 'k:'
+    this.log.debug({ msg: 'acd/storage/redis.heuristicClean picked', key })
+    const exists = await this.store.get(key)
+
+    // if not exists -> found a dangling reference
+    if (exists) {
+      this.log.debug({ msg: 'acd/storage/redis.heuristicClean clear', key })
+      // note: no await
+      this.clearReferences(key)
+    }
+
+    // polling every "interval" until next full sscan iteration
+    this._cleanTimer = setTimeout(() => { this.heuristicClean({ lastCursor: cursor, interval }) }, interval || this.options.invalidation.clear.minInterval)
+  }
+
```

Comment:
It would be helpful to add this script to "clean the cache" as a top-level bin as somebody might want to run it every now and then as a cron.

Reply:
I'll put in doc and examples